provide cloud metrics endpoints, interfaces for providers

anotherchrisberry opened the pull request:
Introducing a new controller, `/cloudMetrics`, to support retrieval of metrics and sets of metric statistics.

New model interfaces:
 * `CloudMetric`
 * `CloudMetricProvider`
 * `CloudMetricStatistics`

I tried to write these interfaces and controller methods as generically as possible, since I do not know what GCP or Azure metrics might look like, and wanted to keep it flexible. I assume it'll be easier to provide more shape to the `CloudMetric`, and then the other interfaces, once they've been implemented.

I wrote this all in Groovy but was told by @ajordens the interfaces and implementations should be in Java, so would love feedback on how to convert some of the groovy bits.

@spinnaker/reviewers PTAL. I would especially like feedback on the right way to handle errors and the shapes of these endpoints and interfaces.

ewiseblatt commented:
I'll look at this tomorrow. Maybe we can talk about it at tomorrow's
meeting. I'll try to read something about Google Cloud Monitoring. I know
it is based on collectd from a third party perspective -- if you arent
using it directly for custom metrics -- to import metrics from 3rd party
services like this.

Does this replace spectator? Why is this in clouddriver? Dont all the
microservices need to support metrics in the same way?

On Wed, Mar 16, 2016 at 11:16 PM, Chris Berry <notifications@github.com>
wrote:

> Introducing a new controller, /cloudMetrics, to support retrieval of
> metrics and sets of metric statistics.
>
> New model interfaces:
>
>    - CloudMetric
>    - CloudMetricProvider
>    - CloudMetricStatistics
>
> I tried to write these interfaces and controller methods as generically as
> possible, since I do not know what GCP or Azure metrics might look like,
> and wanted to keep it flexible. I assume it'll be easier to provide more
> shape to the CloudMetric, and then the other interfaces, once they've
> been implemented.
>
> I wrote this all in Groovy but was told by @ajordens
> <https://github.com/ajordens> the interfaces and implementations should
> be in Java, so would love feedback on how to convert some of the groovy
> bits.
>
> @spinnaker/reviewers <https://github.com/orgs/spinnaker/teams/reviewers>
> PTAL. I would especially like feedback on the right way to handle errors
> and the shapes of these endpoints and interfaces.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/spinnaker/clouddriver/pull/439
> Commit Summary
>
>    - provide cloud metrics endpoints, interfaces for providers
>
> File Changes
>
>    - *A*
>    clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonCloudMetric.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-0> (37)
>    - *A*
>    clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonMetricStatistics.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-1> (42)
>    - *A*
>    clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudMetricProvider.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-2> (147)
>    - *A*
>    clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonMetricStatisticsSpec.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-3> (78)
>    - *A*
>    clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudMetricProviderSpec.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-4> (153)
>    - *A*
>    clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetric.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-5> (27)
>    - *A*
>    clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricProvider.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-6> (53)
>    - *A*
>    clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricStatistics.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-7> (44)
>    - *A*
>    clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/NoopCloudMetricProvider.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-8> (40)
>    - *A*
>    clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudMetricController.groovy
>    <https://github.com/spinnaker/clouddriver/pull/439/files#diff-9> (65)
>
> Patch Links:
>
>    - https://github.com/spinnaker/clouddriver/pull/439.patch
>    - https://github.com/spinnaker/clouddriver/pull/439.diff
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/spinnaker/clouddriver/pull/439>
>

anotherchrisberry commented:
@ewiseblatt These endpoints are not meant to gather metrics about Spinnaker's services - they are there to supply metrics about resources in each cloud provider. Most of those metrics are around instances and server groups, but some of them are around other resources, e.g. message queues, databases, etc.

I'm open to changing the names of the endpoints, too, if that's potentially a point of confusion.

The code here is primarily meant to support viewing and editing scaling policies in the UI:

*(editing)*
<img width="898" alt="screen shot 2016-03-16 at 9 56 36 pm" src="https://cloud.githubusercontent.com/assets/73450/13836787/1eeeb0ac-ebc2-11e5-9065-e6f6cde5653f.png">

*(viewing)*
<img width="676" alt="screen shot 2016-03-16 at 9 58 02 pm" src="https://cloud.githubusercontent.com/assets/73450/13836802/3be5cbdc-ebc2-11e5-9c9a-356fe8f6279f.png">

ewiseblatt commented:
chris, sorry I didnt get to this today. I might later tonight. Otherwise
I'll look in the morning.

On Thu, Mar 17, 2016 at 12:59 AM, Chris Berry <notifications@github.com>
wrote:

> @ewiseblatt <https://github.com/ewiseblatt> These endpoints are not meant
> to gather metrics about Spinnaker's services - they are there to supply
> metrics about resources in each cloud provider. Most of those metrics are
> around instances and server groups, but some of them are around other
> resources, e.g. message queues, databases, etc.
>
> I'm open to changing the names of the endpoints, too, if that's
> potentially a point of confusion.
>
> The code here is primarily meant to support viewing and editing scaling
> policies in the UI:
>
> *(editing)*
> [image: screen shot 2016-03-16 at 9 56 36 pm]
> <https://cloud.githubusercontent.com/assets/73450/13836787/1eeeb0ac-ebc2-11e5-9065-e6f6cde5653f.png>
>
> *(viewing)*
> [image: screen shot 2016-03-16 at 9 58 02 pm]
> <https://cloud.githubusercontent.com/assets/73450/13836802/3be5cbdc-ebc2-11e5-9c9a-356fe8f6279f.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/spinnaker/clouddriver/pull/439#issuecomment-197703650>
>

ewiseblatt commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricStatistics.groovy:
```diff
@@ -0,0 +1,44 @@
+  /**
+   * Unit of measurement for all datapoints; should be the plural form of the unit if applicable, e.g. "bytes"
+   */
+  String unit
```
Is this just a display name, or are these standardized to have semantic meaning? Since you say plural form, I suspect these are just display names.

anotherchrisberry replied:
Correct - it is just a display name. I'm open to a better name for this field to convey its purpose.

ewiseblatt commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricStatistics.groovy:
```diff
@@ -0,0 +1,44 @@
+    Date timestamp
+    Double average
+    Double sum
+    Double sampleCount
```
More FYI than anything else.
I'm not too familiar with Google Cloud Monitoring yet.

Browsing the API, it looks like the model is a bit different.
Points in a time series (which looks like this corresponds to) have a start/end time rather than instantaneous timestamp. Values are not necessarily Double -- they could be any primitive type (string, bool, int64, double) as well as distributions. For purposes of spinnakers, needs double might be sufficient (certainly ok as a starting point). However this interface is restrictive to which metrics you can tap into.

Distributions are bucketed value ranges containing bounds and a count (bounds are Double). So looks like they can comply to this interface. Except there is an overflow and underflow bucket, so I guess those values would get thrown away. Not sure how that plays in practice. We'd have to compute the average (which is reasonable) and leave other statistical possibilities on the table for now.

ewiseblatt commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricProvider.groovy:
```diff
@@ -0,0 +1,53 @@
+   * @param filters a collection of identifiers used to uniquely identify a metric
+   * @return a CloudMetricStatistics object, describing the statistics with timestamps
+   */
+  CloudMetricStatistics getStatistics(String account, String region, Map<String, String> filters)
```
Do you want a time bound here? Probably dont want from the beginning of time and are interested in only some span of recent history. Perhaps this should have a start time or relative time offset from now (e.g. last 5 minutes).

anotherchrisberry replied:
Requiring time bound makes sense here. I've included it in the AWS implementation by looking for `startTime` and `endTime` in the `filters`, but you're right - those should always be included, and should be pushed up the call chain to make the API more understandable. Will update.

ewiseblatt commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricProvider.groovy:
```diff
@@ -0,0 +1,53 @@
+   * @param filters a collection of identifiers used to select a subset of all metrics in the account and region
+   * @return a list of metrics matching the filters
+   */
+  List<T> listMetrics(String account, String region, Map<String, String> filters)
```
Perhaps time bound here too

anotherchrisberry replied:
I'm having trouble coming up with a clear vocabulary. In AWS, a "metric" does include any data points - just a unique identifier, which is a composite key of two fields (namespace, metricName) and one or more dimensions (e.g. AutoScalingGroupName). You have to dive into "metric statistics" to get any actual data.

From AWS's perspective, you can't filter a set of metrics based on when they were collected, so it would not support time bounds here, unfortunately.

ewiseblatt replied:
Does the consumer filter? If so you can do that on their behalf. If not, how do you know when you are starting to surge or decay? Considering time range seems fundamental to making auto scaling decisions, unless you are only dealing with point data, and point data is probably not that useful because [I imagine] you want trends not impulses.

ewiseblatt replied:
Metric being an abstraction without data is fine. That's how GCM works too.

A "Metric" is called a "MetricDescriptor".
https://developers.google.com/resources/api-libraries/documentation/cloudmonitoring/v2beta2/java/latest/com/google/api/services/cloudmonitoring/model/MetricDescriptor.html

You get the data as a Timeseries
https://developers.google.com/resources/api-libraries/documentation/cloudmonitoring/v2beta2/java/latest/com/google/api/services/cloudmonitoring/model/Timeseries.html

Which contains Points
https://developers.google.com/resources/api-libraries/documentation/cloudmonitoring/v2beta2/java/latest/com/google/api/services/cloudmonitoring/model/Point.html

or PointDistributions
https://developers.google.com/resources/api-libraries/documentation/cloudmonitoring/v2beta2/java/latest/com/google/api/services/cloudmonitoring/model/PointDistribution.html

This is pretty much what you have data-wise. You are missing the way to convey what the data represents. That's where the MetricDescription comes in. Minimally you need a label to display and a type so that you can interpret. I think you also need a time range so that consumers can determine rates of interest.

anotherchrisberry replied:
Good point. I'll include time boundaries here.

anotherchrisberry replied:
I've renamed this interface `CloudMetricDescriptor`.

ewiseblatt commented on clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudMetricController.groovy:
```diff
@@ -0,0 +1,65 @@
+  List<CloudMetricProvider> metricProviders
+
+  @RequestMapping(method = RequestMethod.GET, value = "/{cloudProvider}/{account}/{region}")
+  List<CloudMetric> findAll(@PathVariable String cloudProvider,
```
I dont think google metrics are region based, though they support labels so maybe labels are used for regions. I assume this is to support deck. How does time work?

anotherchrisberry replied:
It's to support Deck, but also to support AWS, since we need to create different API clients per region to query the metrics data. I'm assuming it's still desirable on the GCP as a filter, since you can have multiple server groups with the same name (different identifiers, but API users don't want to know those) in different regions.

ewiseblatt commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetric.groovy:
```diff
@@ -0,0 +1,27 @@
+ * supplies "metricName" and "namespace" fields, as well as a collection of "dimensions"
+ */
+interface CloudMetric {
+
```
Who consumes this interface? It seems you are creating provider-specific metrics, which means consumers have to know which provider they are using to do anything. That seems to be in conflict (though not entirely) with having the other interfaces for the values.

I'd think a metric would at least have a name, and perhaps a type. What you call the unit below in the CloudMetricStatistics would instead be "metric" and an instance of this. Then if you have some standard types, any consumer can use the metric interface and could render it (having the name) and have a chance at interpreting it (from the type) without having to know the provider. Nonstandard types would still need additional knowledge to interpret (as it is now) but you'd know it is non-standard for whatever that is worth.

I think a type is a unit, but am not sure what units are. Are these things like "bytes" and "seconds" or could they also be rates ("bytes/second"). I think rates are usually introduced higher in the consumption chain.

anotherchrisberry replied:
This interface is consumed by the controller via the `getMetric` and `listMetrics` methods. The primary purpose of those endpoints (for now, at least) is to supply data to help the user set up autoscaling policies in the UI.

You're right - metrics should at least have a `name`. I will add that to the interface. I also agree that `type`/`unit` would be nice. For whatever reason, AWS doesn't supply the unit of measurement except in the context of an aggregated data point, so determining what the unit should be is non-trivial - you'd have to get data from a time window, and it's possible there's no data, in which case you can't determine it at all. So I'm leaving it off.

To answer your last question - units are all three of those things. These are the possible values AWS will send in the response to getStatistics:

> Seconds, Microseconds, Milliseconds, Bytes, Kilobytes, Megabytes, Gigabytes, Terabytes, Bits, Kilobits, Megabits, Gigabits, Terabits, Percent, Count, Bytes/Second, Kilobytes/Second, Megabytes/Second, Gigabytes/Second, Terabytes/Second, Bits/Second, Kilobits/Second, Megabits/Second, Gigabits/Second, Terabits/Second, Count/Second, None

ewiseblatt commented:
My comments are gut feeling and may very well feel differently once I try to use it or see how it is consumed. My only real concern is how time fits into the interface.

anotherchrisberry commented:
@ewiseblatt I've updated this quite a bit - mostly around the `Metric` interface (renamed to `MetricDescriptor`, added a `name` field), but also making time boundaries part of the API where appropriate and providing defaults via the controller method to get a set of metric data.

ewiseblatt commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricStatistics.groovy:
```diff
@@ -0,0 +1,45 @@
+   */
+  List<CloudMetricDatapoint> datapoints
+
+  @Immutable
```
I wonder if this should be an interface rather than a data object. I am assuming performance doesnt matter. An interface could just wrap your AWS data so these methods are straight through. When you construct the object, you just pass the data instance through and bind it to your class that implements the interface and returns the appropriate fields instead of breaking out the fields.

The benefit is now other platforms can come through and possibly add their additional data and have the option to compute some of it on demand rather than eagerly transform it.

Just a thought. We can refactor it later after adding another provider.

anotherchrisberry replied:
That makes a lot more sense than the current implementation - will update.

ewiseblatt commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricProvider.groovy:
```diff
@@ -0,0 +1,58 @@
+   * @param filters a collection of identifiers used to select a subset of all metrics in the account and region
+   * @return a list of metric descriptors matching the filters
+   */
+  List<T> listMetricDescriptors(String account, String region, Map<String, String> filters)
```
How is this used? Is it odd to have a discovery mechanism for metrics without knowing what the types are. Speaks more to the descriptor than this method.

ewiseblatt commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricProvider.groovy:
```diff
@@ -0,0 +1,58 @@
+   * @param endTime an exclusive timestamp to determine the newest datapoint to return
+   * @return a CloudMetricStatistics object, describing the statistics with timestamps
+   */
+  CloudMetricStatistics getStatistics(String account, String region, String name, Map<String, String> filters,
```
Should "name" be a MetricDescriptor? Your discovery method for names returns the metrics, so seems that where the name comes from to begin with -- unless you know it apriori. That might also reduce the footprint of platform specific anomalies in this interface.

If you have the MetricDescriptor, you dont need the other parameters, though your MetricDescriptor implementation may need to remember them.

ewiseblatt replied:
Since getMetricDescriptor does not take a name, it seems getStatistics should take a descriptor and time bounds since it looks like you dont know the name until you already have a descriptor.

anotherchrisberry replied:
That also makes more sense. I will update accordingly.

anotherchrisberry replied:
Well, I have to walk back that last comment. 

It makes sense, but I can't find a clean way to implement it. To get the statistics for AWS, I need the following fields to identify the metric:
* account
* region
* metricName
* namespace
* dimensions (an arbitrary map of values)

Then, I need these fields to determine how to generate the data points:
* statistic (average, sum, maximum, etc.)
* period (number of seconds per sample)
* startTime
* endTime

I'm guessing GCM will have a different set of fields, so I am unsure how to go about this. I tried adding a `from` method to the `CloudMetricProvier` interface, i.e.
```java
T from(Map<String, String> parameters)
```
but got myself into trouble implementing the AWS side around knowing which fields should be dimensions and which fields should be for generating data. Everything I tried ended up muddling concerns between the controller, the interfaces, and the provider implementations.

For now, I'm going to leave this as-is. If you or someone else can find a good way to refactor this once another cloud provider gets in the mix, I am all for it.

ewiseblatt commented on clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudMetricController.groovy:
```diff
@@ -0,0 +1,102 @@
+    filters.remove("startTime")
+    filters.remove("endTime")
+
+    Long start = startTime ?: clock.instant().minus(24, HOURS).toEpochMilli()
```
Maybe this could be a configurable parameter so you can specify the window you want to use here. A day seems pretty long for autoscaling decisions. I would have guessed you'd be using a much smaller window.

On the end time, I wonder if you want it to be open ended to be the most current data you have. This will wind up with a slightly bigger window but means in practice you wont need to strip off data (unless the endTime is projected into the future to achieve the same effect). Otherwise when you say "now" you are looking into the past (by the time the request is processed) rather than the present. Leaving it open ended lets you defer the end time until you are forced (if at all) to apply an actual value.

anotherchrisberry replied:
The only downside to the relative window is: how do I get a non-relative window, e.g. if I have today's data but also want to see the day before, or if I want to see what Friday nights look like over the past month.

We could add a parameter (I can't think of a name for it, maybe `relativeLookbackInMs` or something along those lines), but I think we'd still want the option for start/end times to be specified.

And we'll have to strip any of these parameters out of the `filters`, just because Spring is going to put all the request parameters into that map, even if they're specified as `@RequestParam`s in the method signature.

For now, I'm going to leave this as-is, but, again, am open to refactoring when you get to the Google implementation.

ewiseblatt commented:
I didnt look that closely at the AWS bits.
I'm ok with things as they are if you want since we know we are going to need to refactor things but not exactly sure what or how until I try to apply this. Depending on how much datapoints are used, making it an interface now could make it easier to refactor later and will likely make it easier for other providers to play around, especially to extend the data model.

duftler commented:
What was the end result on this PR?

I read through the thread again, and it seems like @ewiseblatt was fine with merging the PR as-is, keeping in mind that we'll end up doing some refactoring when other providers start to add support for this. Is that right?

ewiseblatt commented:
Yes, though I would encourage considering a couple things now to make it easier to refactor later.

The interface for the data object is something likely we will need or want. It will certainly make it easier for us to get started since our data model does not align with theirs. So it would be easier on us for him to put that interface in now -- if an interface is in fact the right choice here.

I also think requiring an upper time bound might be clumsy to use. Typically you arent interested in a past window, you are interested in a present window, meaning it ends "now". This means, if using absolute times, the window might be a little bigger than what you asked for (because of latency between the time you for the request and it makes its way to the code filtering the data) but I imagine that's ok. If in practice you always ask for a the bounds {now() - window_secs, now()} then an interface that takes just the window size relative to now() would be both simpler and a bit more precise since you can defer the actual time range further down the call chain. Again, this could be changed later, but doing so now will be less refactoring if that's where we think we'll end up.

anotherchrisberry commented:
I updated everything to an interface (CloudMetricProvider, CloudMetricDescriptor, CloudMetricDatapoint) with minimal fields, so that should be easy to implement.

As far as required time bounds, the current implementation allows for most combinations of start/end times to be specified - they're both optional; `startTime` defaults to 24 hours ago, `endTime` to current time (the AWS API explicitly requires both times to be specified when retrieving statistics). So getting that relative window is possible by just sending a `startTime` of 30 minutes ago.

I'll go ahead and merge it in its current state with the expectation that some refactoring will happen when the next provider implements it.
